### PR TITLE
prewrite_improve: abort Prewrite command after Rollback for the same txn

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,all,
      This PR is for [issue_1649](https://github.com/pingcap/tikv/issues/1649). @disksing @zhangjinpeng1987  PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @hhkbp2

**Comment by disksing** [comment]
Please update the PR's title with more details.

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -201,7 +201,13 @@ impl<'a> MvccTxn<'a> {
                         Err(Error::Committed { commit_ts: ts })
                     }
                     // Rollbacked by concurrent transaction.
-                    None => Ok(()),
+                    None => {
```
Comment (disksing) [inline src/storage/mvcc/txn.rs]: `get_txn_commit_ts` also returns None when there is already a Rollback record, we don't have to insert again under this circumstance. Furthermore, we must not insert again if we want to use SingleDelete later. cc @zhangjinpeng1987

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -201,7 +201,13 @@ impl<'a> MvccTxn<'a> {
...
+                        // insert a Rollback to WriteCF when receives Rollback before Prewrite
+                        let write = Write::new(WriteType::Rollback, self.start_ts, None);
```
Comment (siddontang) [inline src/storage/mvcc/txn.rs]:
```
let ts = self.start_ts;
let write = Write::new(WriteType::Rollback, ts, None);       
```

**Comment by siddontang** [comment]
should add test in tests with raft engine.

File: src/storage/mvcc/txn.rs (hunk 1 above)
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: `get_txn_commit_ts` need return more information to indicate the txn is rollbacked or not.

**Comment by siddontang** [comment]
Is it compatibility for the rolling update?

**Comment by AndreMouche** [comment]
For this case In the rolling update, `tikv` should be updated before `tidb`. @siddontang

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -161,16 +161,17 @@ impl<'a> MvccTxn<'a> {
             }
             _ => {
                 return match try!(self.reader.get_txn_commit_ts(key, self.start_ts)) {
-                    // Committed by concurrent transaction.
-                    Some(_) => Ok(()),
-                    // Rollbacked by concurrent transaction.
+                    Some((_, WriteType::Rollback)) |
                     None => {
+                        // Rollbacked by concurrent transaction.
```
Comment (disksing) [inline src/storage/mvcc/txn.rs]: When commit a key that is not locked, it must be committed or rolled back. So we should panic for _None_.

File: src/storage/mvcc/reader.rs [hunk 4]
```diff
@@ -215,15 +215,13 @@ impl<'a> MvccReader<'a> {
         }
     }
 
-    pub fn get_txn_commit_ts(&mut self, key: &Key, start_ts: u64) -> Result<Option<u64>> {
+    pub fn get_txn_commit_ts(&mut self,
```
Comment (disksing) [inline src/storage/mvcc/reader.rs]: Maybe consider changing the function name.

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -193,15 +194,26 @@ impl<'a> MvccTxn<'a> {
...
+                            info!("txn conflict (committed), key:{}, start_ts:{}, commit_ts:{}",
+                                  key,
+                                  self.start_ts,
+                                  ts);
+                            Err(Error::Committed { commit_ts: ts })
+                        }
                     }
                     // Rollbacked by concurrent transaction.
-                    None => Ok(()),
+                    None => {
```
Comment (disksing) [inline src/storage/mvcc/txn.rs]: Staled comment.

File: src/storage/mvcc/txn.rs (hunk 3 above)
Comment (AndreMouche) [inline src/storage/mvcc/txn.rs]: Since it's not related to this `PR`,  I'd like to fix it in a new `PR`. @disksing

**Comment by disksing** [comment]
LGTM

**Comment by AndreMouche** [comment]
@zhangjinpeng1987 @siddontang PTAL

File: src/storage/mvcc/txn.rs [hunk 6]
```diff
@@ -191,17 +192,27 @@ impl<'a> MvccTxn<'a> {
...
+                            info!("txn conflict (committed), key:{}, start_ts:{}, commit_ts:{}",
+                                  key,
+                                  self.start_ts,
+                                  ts);
+                            Err(Error::Committed { commit_ts: ts })
+                        }
+                    }
+                    None => {
+                        let ts = self.start_ts;
+                        // insert a Rollback to WriteCF when receives Rollback before Prewrite
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: before `Commit`

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -160,17 +160,18 @@ impl<'a> MvccTxn<'a> {
                 (lock.lock_type, lock.short_value.take())
             }
             _ => {
-                return match try!(self.reader.get_txn_commit_ts(key, self.start_ts)) {
-                    // Committed by concurrent transaction.
-                    Some(_) => Ok(()),
-                    // Rollbacked by concurrent transaction.
+                return match try!(self.reader.get_txn_commit_info(key, self.start_ts)) {
+                    Some((_, WriteType::Rollback)) |
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: Do we need another Error type to indicate rollbacked?

File: src/storage/mvcc/txn.rs [hunk 8]
```diff
@@ -853,14 +872,20 @@ mod tests {
...
     }
 
     fn must_get_commit_ts_none(engine: &Engine, key: &[u8], start_ts: u64) {
         let snapshot = engine.snapshot(&Context::new()).unwrap();
         let mut reader = MvccReader::new(snapshot.as_ref(), None, true, None);
-        assert!(reader.get_txn_commit_ts(&make_key(key), start_ts).unwrap().is_none());
+        let ret = reader.get_txn_commit_info(&make_key(key), start_ts);
+        if ret.is_ok() {
+            let (_, write_type) = ret.unwrap().unwrap();
+            assert_eq!(write_type, WriteType::Rollback);
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: `ret` maybe Ok(None)

File: src/storage/mvcc/txn.rs [hunk 9]
```diff
@@ -160,17 +160,18 @@ impl<'a> MvccTxn<'a> {
...
             _ => {
-                return match try!(self.reader.get_txn_commit_ts(key, self.start_ts)) {
-                    // Committed by concurrent transaction.
-                    Some(_) => Ok(()),
-                    // Rollbacked by concurrent transaction.
+                return match try!(self.reader.get_txn_commit_info(key, self.start_ts)) {
+                    Some((_, WriteType::Rollback)) |
                     None => {
+                        // Rollbacked by concurrent transaction.
                         info!("txn conflict (lock not found), key:{}, start_ts:{}, commit_ts:{}",
```
Comment (hhkbp2) [inline src/storage/mvcc/txn.rs]: Change this comment to adapt the modification.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/storage/mvcc/txn.rs (hunk 7 above)
Comment (AndreMouche) [inline src/storage/mvcc/txn.rs]: We check commit_info only when lock is not exist. So when lock is not exist, and data had been marked with committed(is rollback or not exist), return `TxnLockNotFound`.

File: src/storage/mvcc/txn.rs [hunk 10]
```diff
@@ -191,17 +193,27 @@ impl<'a> MvccTxn<'a> {
                 }
             }
             _ => {
-                return match try!(self.reader.get_txn_commit_ts(key, self.start_ts)) {
+                return match try!(self.reader.get_txn_commit_info(key, self.start_ts)) {
                     // Already committed by concurrent transaction.
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: Please fix the comment.

File: src/storage/mvcc/txn.rs [hunk 11]
```diff
@@ -165,17 +165,19 @@ impl<'a> MvccTxn<'a> {
...
+                        // TODO:None should not appear
+                        // Rollbacked by concurrent transaction.
                         info!("txn conflict (lock not found), key:{}, start_ts:{}, commit_ts:{}",
                               key,
                               self.start_ts,
                               commit_ts);
                         Err(Error::TxnLockNotFound)
                     }
+                    // Committed by concurrent transaction.
+                    _ => Ok(()),
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/txn.rs]: I think `Some((_, WriteType::Put)) | Some((_, WriteType::Del)) => ` is more clear.

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
LGTM @disksing PTAL again.

**Comment by disksing** [comment]
LGTM.

**Pull request merged**